File: java/111.minimum-depth-of-binary-tree.java
/*
 * @lc app=leetcode id=111 lang=java
 *
 * [111] Minimum Depth of Binary Tree
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) return 0;
        int leftDepth = minDepth(root.left);
        int rightDepth = minDepth(root.right);
        if (rightDepth == 0) return leftDepth + 1;
        if (leftDepth == 0) return rightDepth + 1;
        return Math.min(leftDepth, rightDepth) + 1;
    }
}

File: go/111.minimum-depth-of-binary-tree.go
/*
 * @lc app=leetcode id=111 lang=golang
 *
 * [111] Minimum Depth of Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func minDepth(root *TreeNode) int {
    if root == nil {
		return 0
	}
	d, curcnt, nextcnt := 1, 1, 0
	queue := []*TreeNode{root}
	for {
		for i := 0; i < curcnt; i++ {
			node := queue[i]
			l, r := node.Left, node.Right
			if l == nil && r == nil {
				return d
			}
			if l != nil {
				queue = append(queue, l)
				nextcnt++
			}
			if r != nil {
				queue = append(queue, r)
				nextcnt++
			}
		}
		queue = queue[curcnt:]
		d, curcnt, nextcnt = d+1, nextcnt, 0
	}
}
// @lc code=end

